src/schema/position.rs

```diff
@@ -65,6 +66,123 @@ impl From<InputObjectTypeDefinitionPosition> for TypeDefinitionPosition {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) enum CompositeTypeDefinitionPosition {
+    Object(ObjectTypeDefinitionPosition),
+    Interface(InterfaceTypeDefinitionPosition),
+    Union(UnionTypeDefinitionPosition),
+}
+
+impl CompositeTypeDefinitionPosition {
+    pub(crate) fn type_name(&self) -> &Name {
+        match self {
+            CompositeTypeDefinitionPosition::Object(type_) => &type_.type_name,
+            CompositeTypeDefinitionPosition::Interface(type_) => &type_.type_name,
+            CompositeTypeDefinitionPosition::Union(type_) => &type_.type_name,
+        }
+    }
+}
+
+impl From<ObjectTypeDefinitionPosition> for CompositeTypeDefinitionPosition {
+    fn from(value: ObjectTypeDefinitionPosition) -> Self {
+        CompositeTypeDefinitionPosition::Object(value)
+    }
+}
+
+impl From<InterfaceTypeDefinitionPosition> for CompositeTypeDefinitionPosition {
+    fn from(value: InterfaceTypeDefinitionPosition) -> Self {
+        CompositeTypeDefinitionPosition::Interface(value)
+    }
+}
+
+impl From<UnionTypeDefinitionPosition> for CompositeTypeDefinitionPosition {
+    fn from(value: UnionTypeDefinitionPosition) -> Self {
+        CompositeTypeDefinitionPosition::Union(value)
+    }
+}
+
+impl From<AbstractTypeDefinitionPosition> for CompositeTypeDefinitionPosition {
+    fn from(value: AbstractTypeDefinitionPosition) -> Self {
+        match value {
+            AbstractTypeDefinitionPosition::Interface(value) => value.into(),
+            AbstractTypeDefinitionPosition::Union(value) => value.into(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) enum AbstractTypeDefinitionPosition {
+    Interface(InterfaceTypeDefinitionPosition),
+    Union(UnionTypeDefinitionPosition),
+}
+
+impl AbstractTypeDefinitionPosition {
+    pub(crate) fn type_name(&self) -> &Name {
+        match self {
+            AbstractTypeDefinitionPosition::Interface(type_) => &type_.type_name,
+            AbstractTypeDefinitionPosition::Union(type_) => &type_.type_name,
+        }
+    }
+}
+
+impl From<InterfaceTypeDefinitionPosition> for AbstractTypeDefinitionPosition {
+    fn from(value: InterfaceTypeDefinitionPosition) -> Self {
+        AbstractTypeDefinitionPosition::Interface(value)
+    }
+}
+
+impl From<UnionTypeDefinitionPosition> for AbstractTypeDefinitionPosition {
+    fn from(value: UnionTypeDefinitionPosition) -> Self {
+        AbstractTypeDefinitionPosition::Union(value)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) enum FieldDefinitionPosition {
+    Object(ObjectFieldDefinitionPosition),
+    Interface(InterfaceFieldDefinitionPosition),
+    Union(UnionTypenameFieldDefinitionPosition),
```

Comment:
same question as below -> unions don't have fields so why do we need this? do we track ALL possible fields from all union implementations?

Reply:
Unions have a single field named `__typename`, which we need to be able to represent in query graphs.